testutil/testpolylog/event.go

```diff
@@ -0,0 +1,166 @@
+
+	// NewEventWithLevelFn is called in the test helper to create a new event
+	// at the given level from the given logger. It is useful for decoupling
+	// test helpers from a specific logger implementation.
+	NewEventWithLevelFn func(
+		*testing.T,
+		polylog.Logger,
+		polylog.Level,
+	) polylog.Event
+)
+
+// FnMethodSpy is a mock which implements a #Fn() method that is intended to be
+// used in tests to assert that the function passed to polylog.Event#Func() is
+// called with the expected arg(s).
+type FnMethodSpy struct{ mock.Mock }
+
+// Fn is a mock method which can be asserted on via the mock.Mock API.
+// See: https://pkg.go.dev/github.com/stretchr/testify@v1.8.4/mock#Mock.
+func (m *FnMethodSpy) Fn(event polylog.Event) {
+	m.Called(event)
+}
+
+// EventMethodsTest is a test case for expressing and exercising polylog.Event
+// methods in a concise way.
+type EventMethodsTest struct {
+	Msg                    string
+	MsgFmt                 string
+	MsgFmtArgs             []any
+	Key                    string
+	Value                  any
+	EventMethodName        string
+	ExpectedOutputContains string
+}
+
+// RunEventMethodTests runs a set of tests for a given level.
+func RunEventMethodTests(
+	t *testing.T,
+	level polylog.Level,
+	tests []EventMethodsTest,
+	newLoggerAndOutput NewLoggerAndOutputFn,
+	newEventWithLevel NewEventWithLevelFn,
+	funcMethodEventTypeName string,
+	getExpectedLevelOutputContains func(level polylog.Level) string,
+) {
+	t.Helper()
+
+	// Title-case level string so that it can be used as the name of the
+	// method to call on the logger using reflect and for the sub-test
+	// descriptions.
+	//
+	// TODO_TECHDEBT/TODO_COMMUNITY: `strings.Title()` is deprecated. Follow
+	// migration guidance in godocs: https://pkg.go.dev/strings@go1.21.4#Title.
+	levelMethodStr := strings.Title(level.String())
+
+	for _, tt := range tests {
+		// If the test case does not specify an event method name, use the test
+		// case's key as the event method name. This is done for convenience only.
+		if tt.EventMethodName == "" {
+			tt.EventMethodName = tt.Key
+		}
+
+		var (
+			eventMethodArgs []reflect.Value
+			doneMethodName  string
+			doneMethodArgs  []reflect.Value
+		)
+
+		// Ensure that calls to #Msg(), #Msgf(), and #Send() are mutually exclusive.
+		switch {
+		case tt.Msg != "":
+			// Set up call args for polylog.Event#Msg() if tt.msg is not emtpy.
+			doneMethodName = "Msg"
+			doneMethodArgs = append(doneMethodArgs, reflect.ValueOf(tt.Msg))
+		case tt.MsgFmt != "":
+			// Set up call args for polylog.Event#Msgf() if tt.msgFmt is not emtpy.
+			doneMethodName = "Msgf"
+			doneMethodArgs = append(
+				doneMethodArgs,
+				reflect.ValueOf(tt.MsgFmt),
+				reflect.ValueOf(tt.MsgFmtArgs),
+			)
+		default:
+			// Default to calling polylog.Event#Send() if tt.msg and tt.msgFmt are
+			// both empty.
+			doneMethodName = "Send"
+		}
+
+		// If tt.EventMethodName and tt.Key are both empty, then use the done
```

Comment:
Need to either add lots of comments in the struct or rethink this because the line of thinking isn't clear to me.

Reply:
Now that we also specify the method name explicitly, in addition to your new comments, the codepath is clear and approachable!